File: benchmarks/cross-language/matrix_multiply_numpy/bench.cs
using System;
using MathNet.Numerics.LinearAlgebra;

class Program
{
    // Hand-written C# baseline for the numeric path: MathNet.Numerics dense
    // matrix multiply — the same backend Sharpy's numpy delegates to, so a
    // Spy/C# ratio near 1.0 means the numpy path adds negligible overhead.
    static void Main()
    {
        int size = 256;
        var a = Matrix<double>.Build.Dense(size, size, (i, j) => (i + j) % 7);
        var b = Matrix<double>.Build.Dense(size, size, (i, j) => (i * j + 1) % 11);

        Matrix<double> result = a * b;
        for (int n = 1; n < 200; n++)
        {
            result = a * b;
        }
        Console.WriteLine($"done: {(long)result[0, 0]}");
    }
}

File: benchmarks/cross-language/matrix_multiply_numpy/bench.py
import numpy as np


def main():
    # Numeric-path matrix multiply: numpy's @ (native BLAS) is the idiomatic fast
    # route, the counterpart to the pure-list matrix_multiply benchmark.
    size = 256
    a = np.array([[float((i + j) % 7) for j in range(size)] for i in range(size)])
    b = np.array([[float((i * j + 1) % 11) for j in range(size)] for i in range(size)])

    result = a @ b
    n = 1
    while n < 200:
        result = a @ b
        n = n + 1
    print(f"done: {int(result[0, 0])}")


if __name__ == "__main__":
    main()
